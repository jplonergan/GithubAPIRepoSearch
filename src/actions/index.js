export const getRepos = (repoLanguage, stringQuery) => ({
  type: 'GET_REPOS',
  repoLanguage,
  stringQuery,
});

export const selectRepoLanguage = (repoLanguage) => ({
  type: 'SELECT_LANGUAGE',
  repoLanguage,
});

export const setStringQuery = (stringQuery) => ({
  type: 'SET_STRING_QUERY',
  stringQuery,
});
