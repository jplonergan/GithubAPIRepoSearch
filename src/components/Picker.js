import * as React from 'react';
import PropTypes from 'prop-types';

function Picker({ value, onChange, options }) {
  return (
    <span>
      <select className="slds-input" onChange={(e) => onChange(e.target.value)} value={value}>
        {options.map((option) => (
          <option value={option} key={option}>
            {option}
          </option>
        ))}
      </select>
    </span>
  );
}

Picker.propTypes = {
  options: PropTypes.arrayOf(PropTypes.string.isRequired).isRequired,
  value: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired,
};

export default Picker;
