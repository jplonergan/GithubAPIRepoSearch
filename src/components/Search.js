import React from 'react';
import { connect } from 'react-redux';
import { Button } from '@salesforce/design-system-react';
import { GITHUB_API_LANGUAGE_OPTIONS } from '../constants/constants';
import Picker from './Picker';
import { selectRepoLanguage, getRepos, setStringQuery } from '../actions';

function Search({
  repoLanguage, stringQuery, getRepos, setStringQuery, selectRepoLanguage,
}) {
  const handleSearch = () => getRepos(repoLanguage, stringQuery);
  const handleInputChange = (value) => setStringQuery(value);
  const handlePickerChange = (e) => selectRepoLanguage(e);

  return (
    <div>
      <h1 className="slds-text-heading_large slds-p-around_large">Workday Github Repository Search</h1>
      <div className="slds-grid slds-wrap">
        <div className="slds-col slds-size_1-of-1 slds-small-size_1-of-1 slds-medium-size_4-of-6 slds-large-size_3-of-6 slds-p-around_small">
          <span>
            <input
              id="base-id"
              className="slds-input"
              name="search"
              value={stringQuery}
              placeholder="Search Repository Name/Description"
              onChange={(e) => handleInputChange(e.target.value)}
            />
          </span>
        </div>
        <div className="slds-col slds-size_1-of-1 slds-small-size_1-of-1 slds-medium-size_2-of-6 slds-large-size_2-of-6 slds-p-around_small">
          <span>
            <Picker
              value={repoLanguage}
              onChange={(e) => handlePickerChange(e)}
              options={GITHUB_API_LANGUAGE_OPTIONS}
            />
          </span>
        </div>
        <div className="slds-col slds-size_1-of-1 slds-small-size_1-of-1 slds-medium-size_6-of-6 slds-large-size_1-of-6 slds-p-around_small">
          <span className="slds-float_right">
            <Button label="Search" variant="brand" onClick={handleSearch} />
          </span>
        </div>
      </div>
    </div>
  );
}

const mapDispatchToProps = {
  selectRepoLanguage,
  getRepos,
  setStringQuery,
};

const mapStateToProps = (state) => ({
  repoLanguage: state.repoLanguage,
  stringQuery: state.stringQuery,
});

export default connect(mapStateToProps, mapDispatchToProps)(Search);
