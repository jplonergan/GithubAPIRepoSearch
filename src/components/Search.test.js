import React from 'react';
import { Provider } from 'react-redux';
import { mount, shallow } from 'enzyme';
import configureStore from 'redux-mock-store';
import Search from './Search';
import { selectRepoLanguage, getRepos, setStringQuery } from '../actions';
import Picker from './Picker';
import '../../setupTests';

describe('>>>H O M E --- REACT-REDUX (Mount + wrapping in <Provider>)', () => {
  const initialState = {
    repoLanguage: 'TypeScript',
    stringQuery: '',
  };
  const mockStore = configureStore();
  let store;
  let wrapper;

  beforeEach(() => {
    store = mockStore(initialState);
    wrapper = mount(
      <Provider store={store}>
        <Search />
      </Provider>,
    );
  });

  it('+++ check Input Prop matches with initialState', () => {
    expect(wrapper.find('input').prop('value')).toEqual(initialState.stringQuery);
  });
  it('+++ check Picker Prop matches with initialState', () => {
    expect(wrapper.find(Picker).prop('value')).toEqual(initialState.repoLanguage);
  });

  it('+++ render the component', () => {
    expect(wrapper.find(Search).length).toEqual(1);
  });

  it('+++ check action on dispatching ', () => {
    const action = store.getActions();
    store.dispatch(getRepos());
    store.dispatch(setStringQuery());
    store.dispatch(selectRepoLanguage());

    expect(action[0].type).toBe('GET_REPOS');
    expect(action[1].type).toBe('SET_STRING_QUERY');
    expect(action[2].type).toBe('SELECT_LANGUAGE');
  });

  it('input onChange function should call setStringQuery', () => {
    const setStringQuery = jest.fn();

    const { WrappedComponent } = Search;
    const newwrapper = shallow(<WrappedComponent setStringQuery={setStringQuery} />);
    newwrapper
      .find('input')
      .at(0)
      .simulate('change', { target: { name: 'search', value: 'test' } });
    expect(setStringQuery).toHaveBeenCalled();
  });

  it('should render correctly <Search> component', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
