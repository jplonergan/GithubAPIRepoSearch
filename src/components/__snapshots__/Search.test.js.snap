// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`>>>H O M E --- REACT-REDUX (Mount + wrapping in <Provider>) should render correctly <Search> component 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(Search)>
    <Search
      getRepos={[Function]}
      repoLanguage="TypeScript"
      selectRepoLanguage={[Function]}
      setStringQuery={[Function]}
      stringQuery=""
    >
      <div>
        <h1
          class="slds-text-heading_large slds-p-around_large"
        >
          Workday Github Repository Search
        </h1>
        <div
          className="slds-grid slds-wrap"
        >
          <div
            className="slds-col slds-size_1-of-1 slds-small-size_1-of-1 slds-medium-size_4-of-6 slds-large-size_3-of-6 slds-p-around_small"
          >
            <span>
              <input
                className="slds-input"
                id="base-id"
                name="search"
                onChange={[Function]}
                placeholder="Repository Name"
                value=""
              />
            </span>
          </div>
          <div
            className="slds-col slds-size_1-of-1 slds-small-size_1-of-1 slds-medium-size_2-of-6 slds-large-size_2-of-6 slds-p-around_small"
          >
            <span>
              <Picker
                onChange={[Function]}
                options={
                  Array [
                    "C",
                    "C++",
                    "C#",
                    "GO",
                    "Java",
                    "JavaScript",
                    "PHP",
                    "Python",
                    "Ruby",
                    "Scala",
                    "TypeScript",
                  ]
                }
                value="TypeScript"
              >
                <span>
                  <select
                    className="slds-input"
                    onChange={[Function]}
                    value="TypeScript"
                  >
                    <option
                      key="C"
                      value="C"
                    >
                      C
                    </option>
                    <option
                      key="C++"
                      value="C++"
                    >
                      C++
                    </option>
                    <option
                      key="C#"
                      value="C#"
                    >
                      C#
                    </option>
                    <option
                      key="GO"
                      value="GO"
                    >
                      GO
                    </option>
                    <option
                      key="Java"
                      value="Java"
                    >
                      Java
                    </option>
                    <option
                      key="JavaScript"
                      value="JavaScript"
                    >
                      JavaScript
                    </option>
                    <option
                      key="PHP"
                      value="PHP"
                    >
                      PHP
                    </option>
                    <option
                      key="Python"
                      value="Python"
                    >
                      Python
                    </option>
                    <option
                      key="Ruby"
                      value="Ruby"
                    >
                      Ruby
                    </option>
                    <option
                      key="Scala"
                      value="Scala"
                    >
                      Scala
                    </option>
                    <option
                      key="TypeScript"
                      value="TypeScript"
                    >
                      TypeScript
                    </option>
                  </select>
                </span>
              </Picker>
            </span>
          </div>
          <div
            className="slds-col slds-size_1-of-1 slds-small-size_1-of-1 slds-medium-size_6-of-6 slds-large-size_1-of-6 slds-p-around_small"
          >
            <span
              className="slds-float_right"
            >
              <SLDSButton
                assistiveText={
                  Object {
                    "icon": "",
                  }
                }
                disabled={false}
                hint={false}
                iconSize="medium"
                label="Search"
                onClick={[Function]}
                responsive={false}
                type="button"
                variant="brand"
              >
                <button
                  className="slds-button slds-button_brand"
                  disabled={false}
                  onClick={[Function]}
                  type="button"
                >
                  Search
                </button>
              </SLDSButton>
            </span>
          </div>
        </div>
      </div>
    </Search>
  </Connect(Search)>
</Provider>
`;
